<strong>Дана система линейных алгебраических уравнений с четырьмя неизвестными.
    
A. Преобразовать систему к виду, пригодному для применения принципа сжимающих операторов.

B. Методом простых итераций найти приближенные решения с точностью $10^{-2}$ и с точностью $10^{-4}$, используя априорную и апостериорную оценки числа итераций.

C. Найти точное решение системы и сравнить с приближенными.</strong>

Запишем систему в матричной форме $Αx = b$


In [1]:
import numpy as np
import sympy as sp 
from IPython.display import display

sp.init_printing(use_latex='mathjax')

a = np.array([[7., 7., -7., -2.],
              [3., 4., 5., 8.],
              [2., 2., 2., 1.],
              [2., 0., -2., -1.]])

b = np.array([[5.],
              [60.], 
              [27.], 
              [-1.]])

np.linalg.det(a)

308.0

Поскольку $det(A) = 308 \neq 0$, то система $Ax = b$ имеет единственное решение.

<i>A. Преобразуем систему $Ax = b$ к виду, пригодному для применения принципа сжимающих операторов.</i>



In [2]:
eigenvalues = np.linalg.eigvals(np.dot(a.transpose(), a))
lambda_0 = np.amax(eigenvalues)
display(lambda_0)

158.086803063

Вычислили $\lambda(A^TA)$ - максимальное собственное число матрицы $A^TA$.

In [3]:
i_matrix = np.array([[1., 0., 0., 0.],
              [0., 1., 0., 0.],
              [0., 0., 1., 0.],
              [0., 0., 0., 1.]])

c = np.subtract(i_matrix, (np.dot(a.transpose(), a)/lambda_0))
d = np.dot(a.transpose(), b)/lambda_0

lambda_1 = np.amax(np.linalg.eigvals(c))
display(lambda_1)

0.98884670963

Вычислили $\lambda(C)$ - максимальное собственное число матрицы $C = I - \frac{A^TA}{\lambda(A^TA)}$.

<i>B. Методом простых итераций найдем приближенные решения системы $Cx + d = x$ с точностью $\varepsilon = 10^{-2}$ и $\varepsilon = 10^{-4}$, используя априорную и апостериорную оценки числа итераций.</i>

In [4]:
import math

def dist(x1, x0):
    calculated_vector = np.subtract(x1, x0)
    dist = 0
    for i in np.nditer(calculated_vector):
        dist += i ** 2
    return math.sqrt(dist)

def is_enough(alp, x1, x0, eps):
    return (alp/(1 - alp))*dist(x1, x0) <= eps

def apriori_weight(alp, eps, x0, x1):
    return (math.log(eps*(1 - alp)/dist(x0, x1), alp)) + 1

In [5]:
alpha = lambda_1
epsilon_1 = 10**(-2) 

x_0 = np.array([[2.], 
              [-1.], 
              [3.], 
              [0.]])
display(x_0)

array([[ 2.],
       [-1.],
       [ 3.],
       [ 0.]])

Взяли произвольным образом начальное приближение $x_0$.

Найдем априорную оценку числа итераций $N_{apr}$ при $\varepsilon = 10^{-2}$

In [6]:
x_1 = np.dot(c, x_0) + d
display(apriori_weight(alpha, epsilon_1, x_0, x_1))

927.1854835169232

Следовательно, $N_{apr} = 927$ при $\varepsilon = 10^{-2}$

In [7]:
x_f = x_0
i = 0
while True:
    i += 1
    x_f_prev = x_f
    x_f = np.dot(c, x_f) + d
    if is_enough(alpha, x_f, x_f_prev, epsilon_1):
        break
display(i)
display(x_f)

450

array([[ 5.99283718],
       [ 1.00669114],
       [ 5.99920192],
       [ 0.99977721]])

Фактически же понадобилось 450 итерации и получили приближенное решение.

In [8]:
epsilon_2 = 10**(-4)
x_1 = np.dot(c, x_0) + d
display(apriori_weight(alpha, epsilon_2, x_0, x_1))

1337.7764954024437

Следовательно, $N_{apr} = 1337$ при $\varepsilon = 10^{-4}$

In [9]:
x_s = x_0
i = 0
while True:
    i += 1
    x_s_prev = x_s
    x_s = np.dot(c, x_s) + d
    if is_enough(alpha, x_s, x_s_prev, epsilon_2):
        break
display(i)
display(x_s)

859

array([[ 5.99992984],
       [ 1.00007001],
       [ 5.99999863],
       [ 0.99999094]])

Фактически же понадобилось 859 итерации и получили приближенное решение.

<i>C. Найдем точное решение системы $Ax = b$ и сравним его с приближенными. </i>

In [10]:
x_exact = np.linalg.solve(a, b)

После нахождения точного решения системы, сравним точное решение с приближенными в евклидовой метрике $\rho_2$ пространства $R^{4}$:

In [11]:
dist(x_exact, x_f) < epsilon_1

True

In [12]:
dist(x_exact, x_s) < epsilon_2

True

Таким образом, точность приближенных решений соответствует поставленной задаче.